src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -0,0 +1,1305 @@
+
+            if (forkTracker)
+            {
+                newTrackerMap = newTrackerMap.Add(projectId, tracker.Fork(newProjectState, translate));
+            }
+        }
+
+        return this.Branch(
+            newSolutionState,
+            projectIdToTrackerMap: newTrackerMap);
+    }
+
+    private ImmutableDictionary<ProjectId, ICompilationTracker> CreateCompilationTrackerMap(ProjectId changedProjectId, ProjectDependencyGraph dependencyGraph)
+    {
+        if (_projectIdToTrackerMap.Count == 0)
+            return _projectIdToTrackerMap;
+
+        using var _ = ArrayBuilder<KeyValuePair<ProjectId, ICompilationTracker>>.GetInstance(_projectIdToTrackerMap.Count, out var newTrackerInfo);
+        var allReused = true;
+        foreach (var (id, tracker) in _projectIdToTrackerMap)
+        {
+            var localTracker = tracker;
+            if (!CanReuse(id))
+            {
+                localTracker = tracker.Fork(tracker.ProjectState, translate: null);
+                allReused = false;
+            }
+
+            newTrackerInfo.Add(new KeyValuePair<ProjectId, ICompilationTracker>(id, localTracker));
+        }
+
+        if (allReused)
+            return _projectIdToTrackerMap;
+
+        return ImmutableDictionary.CreateRange(newTrackerInfo);
+
+        // Returns true if 'tracker' can be reused for project 'id'
+        bool CanReuse(ProjectId id)
+        {
+            if (id == changedProjectId)
+            {
+                return true;
+            }
+
+            return !dependencyGraph.DoesProjectTransitivelyDependOnProject(id, changedProjectId);
+        }
+    }
+
+    /// <inheritdoc cref="SolutionState.AddProject(ProjectInfo)"/>
+    public SolutionCompilationState AddProject(ProjectInfo projectInfo)
+    {
+        var newSolutionState = this.Solution.AddProject(projectInfo);
+        var newTrackerMap = CreateCompilationTrackerMap(projectInfo.Id, newSolutionState.GetProjectDependencyGraph());
+
+        return Branch(
+            newSolutionState,
+            projectIdToTrackerMap: newTrackerMap);
+    }
+
+    /// <inheritdoc cref="SolutionState.RemoveProject(ProjectId)"/>
+    public SolutionCompilationState RemoveProject(ProjectId projectId)
+    {
+        var newSolutionState = this.Solution.RemoveProject(projectId);
+        var newTrackerMap = CreateCompilationTrackerMap(projectId, newSolutionState.GetProjectDependencyGraph());
+
+        return this.Branch(
+            newSolutionState,
+            projectIdToTrackerMap: newTrackerMap.Remove(projectId));
+    }
+
+    /// <inheritdoc cref="SolutionState.WithProjectAssemblyName"/>
+    public SolutionCompilationState WithProjectAssemblyName(
+        ProjectId projectId, string assemblyName)
+    {
+        return ForkProject(
+            this.Solution.WithProjectAssemblyName(projectId, assemblyName),
+            new CompilationAndGeneratorDriverTranslationAction.ProjectAssemblyNameAction(assemblyName),
+            forkTracker: true);
```

Comment:
most methods are simple like this.  they call into SolutionState to fork itself, and they passed the forked info into ForkProject to decide waht to do with it.  If the underlying solution didn't change, this will bail out quick, returning 'this' object.

One downside here is that we are always doing the allocation of the action.  In practice this shouldn't be an issue as most changes *do* change the state.

If this *does* bcome an allocation we want to avoid, we can pass a static lammbda here in the future.